specifyweb/frontend/js_src/lib/components/SchemaViewer/TableList.tsx

```diff
@@ -42,46 +42,90 @@ export function SchemaViewerTableList<
+  );
+  return (
+    <GenericSortedDataViewer<DATA>
+      cellClassName={() =>
+        'border border-gray-400 p-2 dark:border-neutral-500 print:p-1'
+      }
+      className="w-fit border border-gray-400 dark:border-neutral-500"
+      data={data}
+      getLink={getLink}
+      headerClassName="border"
+      headerElements={headerElements}
+      headers={headers}
+    />
+  );
+}
+
+export function GenericSortedDataViewer<
+  DATA extends SchemaViewerRow<RR<string, SchemaViewerValue>>
+>({
+  headers,
+  data,
+  getLink,
+  headerElements,
+  className = '',
+  headerClassName = '',
+  cellClassName,
+}: {
+  readonly headers: RR<string, JSX.Element | LocalizedString>;
+  readonly headerElements: RA<JSX.Element>;
+  readonly headerClassName?: string;
+  readonly className?: string;
+  readonly data: RA<DATA>;
+  readonly getLink: ((row: DATA) => string) | undefined;
+  readonly cellClassName?: (
+    row: DATA,
+    column: keyof DATA,
+    index: number
+  ) => string;
+}): JSX.Element {
+  const indexColumn = Object.keys(headers)[0];
+
   return (
     <div
-      className={`
-        grid-table
-        w-fit flex-1 grid-cols-[repeat(var(--cols),auto)] rounded border border-gray-400 dark:border-neutral-500 print:p-1
-        ${className}
-      `}
+      className={`grid-table flex-1 grid-cols-[repeat(var(--cols),auto)] rounded print:p-1 ${className}`}
       role="table"
       style={{ '--cols': Object.keys(headers).length } as React.CSSProperties}
     >
       <div role="row">
-        {Object.entries(headers).map(([name, label]) => (
+        {headerElements.map((element, index) => (
           <div
             className={`
-              sticky top-0 border border-gray-400 bg-[color:var(--background)]
+              sticky top-0 border-gray-400 bg-[color:var(--background)]
               p-2 font-bold dark:border-neutral-500 print:p-1
-            `}
-            key={name}
+            ${headerClassName}`}
+            key={index}
             role="columnheader"
           >
-            <Button.LikeLink
-              onClick={(): void => handleSort(name as FIELD_NAME)}
-            >
-              {label}
-              <SortIndicator fieldName={name} sortConfig={sortConfig} />
-            </Button.LikeLink>
+            {element}
           </div>
         ))}
       </div>
       <div role="rowgroup">
-        {data.map((row) => {
+        {data.map((row, index) => {
           const children = Object.keys(headers).map((column) => {
             const data = row[column];
             return (
-              <Cell key={column}>
+              <Cell
+                className={cellClassName?.(row, column, index)}
+                key={column}
+              >
                 {Array.isArray(data) ? data[1] : row[column]}
               </Cell>
             );
           });
-          const key = row[indexColumn]?.toString();
+          const key = `${row[indexColumn]?.toString()}-${index}`;
```

Comment:
I don't know how safe it is to assume that first column is a string
plus, the fact that you also used index in the key makes it redundant - index is always unique, so it's needless to prepend `row[indexColumn]?.toString()`.

could you remove index from this string (as it defeats the purpose), but then throw an exception here if `process.env.NODE_ENV === 'development' && typeof row[indexColumn] !== 'string'`?

the reason key is needed is so that if items are sorted in the column, React can see that and reorder the DOM elements accordingly, rather than mutate each row element to match what new row element is expected at that position. but if you use `index` as your key, that all doesn't work. that's why it's a react best-practice to not use `index` if you expect your items could ever be reordered

cc @CarolineDenis

Reply:
wouldn't index column be always the same? It takes in the first value, that's why. Actually, just putting key=column would work, because column is a key of object. js would enforce uniqueness

Reply:
nvm I misunderstood. Used your solution